File: api/src/main/java/fr/huiitre/tools/modules/core/user/application/usecase/ListUsersUseCase.java
package fr.huiitre.tools.modules.core.user.application.usecase;

import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.core.user.application.ports.UserRepository;
import fr.huiitre.tools.modules.core.user.application.view.UserAdminView;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ListUsersUseCase implements SecuredUseCase {

    @Override
    public RoleCode requiredRole() {
        return RoleCode.ADMIN;
    }

    private final UserRepository userRepository;

    public ListUsersUseCase(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public List<UserAdminView> execute() {
        return userRepository.findAllForAdmin();
    }
}

File: api/Modules/Core/Users/Application/Usecases/ListUsersUseCase.cs
using Tools.Api.Modules.Core.Security.Application.Ports;
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;
using Tools.Api.Modules.Core.Users.Application.Dto;

namespace Tools.Api.Modules.Core.Users.Application.Usecases;

// Cas d'usage administrateur : lister les utilisateurs pour le tableau d'administration.
//
// Rôle minimum ADMIN, comme l'API Java. À noter que ADMIN est plus élevé que TECH dans la
// hiérarchie (voir RoleCode) : un compte TECH n'a donc pas accès à cette liste.
public sealed class ListUsersUseCase(
    UseCaseAuthorizer authorizer,
    IUserRepository userRepository
) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.Admin;

    public Task<IReadOnlyList<UserAdminDto>> Execute()
    {
        return userRepository.FindAllForAdminAsync();
    }
}
